### DISCUSSION TITLE
Error building Bromite as part of an AOSP build - Any help would be appreciated!

### DISCUSSION THREAD

**Discussion by Bungeetaco** [bromite/bromite#1399] [Q&A] [answered]
Hi there!

First of all, thank you for creating this project and maintaining it! Chromium is a great piece of software not considering the privacy blunders it originally comes with, so I am quite a fan of this project as it aims to fix these flaws.

I am opening this discussion to request more information regarding an issue I have encountered attempting to build the latest Bromite build via the [CHAOSP](https://github.com/CaseyBakey/chaosp) build script created by CaseyBakey for my Pixel device. This is a question in line with issue #1354 that was opened last month: [Race condition crash during build of Bromite (related to AdBlockUpdaterService patch)](https://github.com/bromite/bromite/issues/1354). I couldn't find a follow-up discussion on the subject, which is why I am posting this as I would like to know how this build error could be fixed if possible.

When building Bromite as part of my AOSP build, the build seems to fail at the following:
```
FAILED: android_clang_arm/obj/chrome/browser/browser_process/browser_process.o 
../../third_party/llvm-build/Release+Asserts/bin/clang++ -MMD -MF android_clang_arm/obj/chrome/browser/browser_process/browser_process.o.d -DOFFICIAL_BUILD -D_GNU_SOURCE -DANDROID -DHAVE_SYS_UIO_H -DANDROID_NDK_VERSION_ROLL=r22_1 -DCR_CLANG_REVISION=\"llvmorg-13-init-10392-gd3676d4b-2\" -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D_FORTIFY_SOURCE=2 -D_LIBCPP_ABI_UNSTABLE -D_LIBCPP_DISABLE_VISIBILITY_ANNOTATIONS -D_LIBCXXABI_DISABLE_VISIBILITY_ANNOTATIONS -D_LIBCPP_ENABLE_NODISCARD -D_LIBCPP_HAS_NO_VENDOR_AVAILABILITY_ANNOTATIONS -DCR_LIBCXX_REVISION=8fa87946779682841e21e2da977eccfb6cb3bded -DNDEBUG -DNVALGRIND -DDYNAMIC_ANNOTATIONS_ENABLED=0 -I../.. -Iandroid_clang_arm/gen -I../../third_party/perfetto/include -Iandroid_clang_arm/gen/third_party/perfetto/build_config -Iandroid_clang_arm/gen/third_party/perfetto -I../../third_party/abseil-cpp -I../../third_party/boringssl/src/include -I../../third_party/protobuf/src -Iandroid_clang_arm/gen/protoc_out -fprofile-sample-use=../../chrome/android/profiles/afdo.prof -fprofile-sample-accurate -fno-delete-null-pointer-checks -fno-ident -fno-strict-aliasing -fstack-protector-strong -fwrapv -funwind-tables -fPIC -fcolor-diagnostics -fmerge-all-constants -fcrash-diagnostics-dir=../../tools/clang/crashreports -mllvm -instcombine-lower-dbg-declare=0 -flto=thin -fsplit-lto-unit -fcomplete-member-pointers -ffunction-sections -fno-short-enums --target=arm-linux-androideabi21 -march=armv7-a -mfloat-abi=softfp -mtune=generic-armv7-a -Xclang -fdebug-compilation-dir -Xclang . -no-canonical-prefixes -mfpu=neon -mthumb -Wall -Werror -Wextra -Wimplicit-fallthrough -Wunreachable-code -Wthread-safety -Wextra-semi -Wno-missing-field-initializers -Wno-unused-parameter -Wno-c++11-narrowing -Wno-unneeded-internal-declaration -Wno-undefined-var-template -Wno-psabi -Wno-ignored-pragma-optimize -Wno-implicit-int-float-conversion -Wno-final-dtor-non-final-class -Wno-builtin-assume-aligned-alignment -Wno-deprecated-copy -Wno-non-c-typedef-for-linkage -Wno-max-tokens -Oz -fdata-sections -ffunction-sections -fomit-frame-pointer -g1 -fdebug-info-for-profiling -fvisibility=hidden -Xclang -add-plugin -Xclang find-bad-constructs -Xclang -plugin-arg-find-bad-constructs -Xclang checked-ptr-as-trivial-member -Xclang -plugin-arg-find-bad-constructs -Xclang check-ipc -Wheader-hygiene -Wstring-conversion -Wtautological-overlap-compare -DPROTOBUF_ALLOW_DEPRECATED=1 -std=c++14 -fno-trigraphs -Wno-trigraphs -fno-exceptions -fno-rtti -nostdinc++ -isystem../../buildtools/third_party/libc++/trunk/include -isystem../../buildtools/third_party/libc++abi/trunk/include --sysroot=../../third_party/android_ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot -fvisibility-inlines-hidden -c ../../chrome/browser/browser_process.cc -o android_clang_arm/obj/chrome/browser/browser_process/browser_process.o
In file included from ../../chrome/browser/browser_process.cc:5:
In file included from ../../chrome/browser/browser_process.h:25:
In file included from ../../components/component_updater/adblock_updater_service.h:34:
In file included from ../../components/component_updater/download_filters_task.h:31:
../../services/network/public/cpp/shared_url_loader_factory.h:14:10: fatal error: 'services/network/public/mojom/url_loader_factory.mojom.h' file not found
#include "services/network/public/mojom/url_loader_factory.mojom.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
[56979/106672] CXX android_clang_arm/obj/chrome/browser/p...nd/mojo_bindings/promo_browser_command.mojom-test-utils.o
ninja: build stopped: subcommand failed. 
```

The compiler seems to look for a file named 'services/network/public/mojom/url_loader_factory.mojom.h' but fails to find it, which seems to mean that the file is not being generated as it should as part of such a build.

I'm not a programmer, so attempting to fix this is a bit of an exercise in itself for me. From looking at the issue mentioned above that was opened by @ConradGyllenhaal, they state that to them, it looks like it might be a matter of adding "url_loader_base" in the deps section of the "component_updater" static_library section of ./components/component_updater/BUILD.gn, but I am not sure how one would go about doing that, or if this is actually the issue.

Would anyone here perhaps be able to help in figuring this out, so that this problem could be fixed for users who build from source and continue including Bromite into image builds? 

Any assistance would be highly appreciated!

**Comment by uazo** [comment, marked as answer]
> Would anyone here perhaps be able to help in figuring this out,

it is strange, I do not detect that problem in build but I use goma and perhaps the resulting build order is different.

> it looks like it might be a matter of adding "url_loader_base" in the deps section of the "component_updater" 

is correct, probably needs to be added "//services/network/public/mojom" in the deps section
you could also try with `gn check`

**Reply by Bungeetaco** [reply]
I'll try that to check if it makes a difference and report back. Thank you very much!

**Reply by Bungeetaco** [reply]
Doing this seems to have resolved the build issue, and I was able to build Bromite from source after making this change using the following patch after bromite's patches during build time:
```
diff --git a/components/component_updater/BUILD.gn b/components/component_updater/BUILD.gn
index cdd4003..96a4ed4 100644
--- a/components/component_updater/BUILD.gn
+++ b/components/component_updater/BUILD.gn
@@ -36,6 +36,7 @@ static_library("component_updater") {
     "//components/crx_file",
     "//components/update_client",
     "//components/version_info",
+    "//services/network/public/mojom",
     "//third_party/boringssl:boringssl",
     "//ui/base",
     "//url",
```
Of note is that the Bromite-AdBlockUpdaterService.patch file already edits this file, so this patch needs to be applied after rather than before. Wondering if there would be an impact adding this line to the file directly to the project, or if this would cause issues with other build methods, seeing as that would solve that problem for other users without having to apply an extra patch perhaps.

Thanks a lot for the quick assistance, I really appreciate it!

**Reply by uazo** [reply]
there is no problem adding that line to the patch.

by the way @csagan5 that patch should be modified, because it currently breaks the execution of the unit tests by crashing (it seems that `ThreadChecker` does not work in the tests on the emulator). let's leave a few days and I'll show you: I'm starting to put some tests for bromite in my patches

**Reply by csagan5** [reply]
I will add that change to next release; as for other changes related to tests: patches are welcome.
